### PR TITLE
Implemented generic type substitution for tuple types.

### PR DESCRIPTION
**Pull request by VSadov**

### CODE REVIEW

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - please review

File: src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs [hunk 1]
```diff
@@ -47,6 +47,15 @@ internal NamedTypeSymbol SubstituteNamedType(NamedTypeSymbol previous)
                 return (oldFieldTypes == newFieldTypes) ? previous : AnonymousTypeManager.ConstructAnonymousTypeSymbol(previous, newFieldTypes);
             }
 
+            if (previous.IsTupleType)
+            {
+                TupleTypeSymbol previousTuple = (TupleTypeSymbol)previous;
+                NamedTypeSymbol oldUnderlyingType = previousTuple.UnderlyingTupleType;
+                NamedTypeSymbol newUnderlyingType = (NamedTypeSymbol)SubstituteType(previousTuple.UnderlyingTupleType).Type;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs]: ```(NamedTypeSymbol)SubstituteType(oldUnderlyingType).Type```?

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs (hunk 1 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs]: right. I considered removing the temp here, but i need it anyways, so why not to use it here.

**Pull request merged**